server/lib/parties/party-service.ts

```diff
@@ -280,6 +282,48 @@ export default class PartyService {
     })
   }
 
+  kickPlayer(partyId: string, kickingUser: PartyUser, target: PartyUser, time = Date.now()) {
+    const party = this.parties.get(partyId)
+    if (!party || !party.members.has(kickingUser.id)) {
+      throw new PartyServiceError(
+        PartyServiceErrorCode.NotFoundOrNotInParty,
+        "Party not found or you're not in it",
+      )
+    }
+
+    if (kickingUser.id !== party.leader.id) {
+      throw new PartyServiceError(
+        PartyServiceErrorCode.InsufficientPermissions,
+        'Only party leaders can kick other people',
+      )
+    }
+
+    if (!party.members.has(target.id)) {
+      throw new PartyServiceError(
+        PartyServiceErrorCode.InvalidAction,
+        "Can't kick player who is not in your party",
+      )
+    }
+
+    if (kickingUser.id === target.id) {
+      throw new PartyServiceError(PartyServiceErrorCode.InvalidAction, "Can't kick yourself")
+    }
+
+    party.members.delete(target.id)
```

Comment:
Feels kinda bad that this can't reuse `removeClientFromParty`, I wonder if instead you could separate the two pieces here:

1) A `'kick'` action that notifies the party that someone is being kicked
2) A `'leave'` action that notifies the party that this person is now gone

And then you could just publish this first action, followed by a call to `removeClientFromParty` to do all the cleanup

Probably needs a bit more thinking on the client as far as showing messages and stuff, but seems doable (on the client the `kick` could update the members + add a message, and then `leave` could only add a message if the person was in `members`? That or just start managing those messages entirely on the server)

Reply:
Yeah, I was trying to avoid the situation we had in the lobby code where we had to pass "removal type" through several methods to do the right thing. And yeah, could split this into two pieces with one piece having a shared logic that removes the player from the party. The reason why I didn't do that here is because the two features of removing a player from the party are "changing the host if it leaves" and "removing the party if last person leaves"; both of which are not applicable here since you can't kick yourself from a party. But I guess it doesn't hurt having that logic even when cleaning the kicked player, especially since we might add additional logic that needs to be shared between these two scenarios.

Reply:
It definitely does not hurt to run logic that won't change anything, yes. Running (cheap) things unnecessarily is far, far, far preferable to having logic copy-pasted a bunch of places. We (I) have already caused bugs in the past from not properly tracking down all the places lobbies do certain things that would ideally be shared, and the easy way to prevent that is to share it.

Optimize for stuff that will keep working when we need to add or change the implementation